### PR TITLE
test: improve test coverage for PR approval attestation (#466)

### PR DESCRIPTION
**Pull request by DevAnuragT**
This PR addresses issue #466 by enhancing the test coverage for GitHub Pull Request approval attestations. The changes include:

- Adding tests for edge cases involving empty or missing approvers and dismissedApprovers fields, including cases where one or both are nil or empty.

- Covering retrieval and indexing logic by asserting that attestations fetched by review ID and related info match the stored data.

- Validating the overall structure of generated predicates by comparing envelopes to ensure correctness.

- No external GitHub API calls are mocked as they are not involved in the tested code paths.

### CODE REVIEW

File: internal/attestations/github_test.go [hunk 1]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+			}
+		}
+		if at > 0 {
+			result[i] = email[:at]
+		} else {
+			result[i] = email
+		}
+	}
+	return result
+}
```
Comment (patzielinski) [inline internal/attestations/github_test.go:157]: I don't think this is necessary, we should test for the exact data unless it's problematic to do so.

File: internal/attestations/github_test.go [hunk 2]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+	require.NoError(t, err)
+
+	// Now unmarshal the Predicate from structpb.Struct into the custom Go struct
+	var parsed v01.PullRequestApprovalAttestation
+	err = common.UnmarshalPBStruct(statement.Predicate, &parsed)
+	require.NoError(t, err)
+
+	// Assertions
+	assert.ElementsMatch(t, []string{"alice", "bob"}, stripEmail(parsed.GetApprovers()))
+	assert.ElementsMatch(t, []string{"charlie"}, stripEmail(parsed.GetDismissedApprovers()))
```
Comment (patzielinski) [inline internal/attestations/github_test.go:136]: I think an `assert.Equal` would be preferred here, asserting that each element is equal, as well as an assertion that the length is what is expected.

File: internal/attestations/github_test.go [hunk 3]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+		}
+	}
+	return result
+}
+
+/*
+TestNewPullRequestApprovalAttestation_ErrorOnEmptyLists verifies that
+NewGitHubPullRequestApprovalAttestation returns an error when both
+approvers and dismissedApprovers are empty.
+*/
```
Comment (patzielinski) [inline internal/attestations/github_test.go:163]: When these are moved into `t.Run` functions, you can copy the description inside the `t.Run` for this test.

File: internal/attestations/github_test.go [hunk 4]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+	ref := "refs/heads/main"
+	fromID := gitinterface.ZeroHash.String()
+	treeID := gitinterface.ZeroHash.String()
+
+	// Create an attestation with no approvers or dismissed approvers
+	attestation, err := NewGitHubPullRequestApprovalAttestation(ref, fromID, treeID, nil, nil)
+
+	// The attestation should be nil and an error should be returned
+	assert.Nil(t, attestation)
+	assert.Error(t, err)
```
Comment (patzielinski) [inline internal/attestations/github_test.go:174]:
```suggestion
	assert.ErrorIs(t, err)
```

File: internal/attestations/github_test.go [hunk 5]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+	toID := gitinterface.ZeroHash.String()
+	approvers := []string{"alice@example.com"}
+	dismissedApprovers := []string{"bob@example.com"}
+
+	// Create an attestation envelope with both approvers and dismissedApprovers
+	env := createGitHubPullRequestApprovalAttestationEnvelope(t, ref, fromID, toID, approvers, dismissedApprovers)
+
+	// Verify that the envelope and payload are valid
+	assert.NotNil(t, env)
+	assert.NotEmpty(t, env.Payload)
```
Comment (patzielinski) [inline internal/attestations/github_test.go:232]: These should be compared against an expected envelope.

File: internal/attestations/github_test.go [hunk 6]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+}
+
+// createGitHubPullRequestApprovalAttestationEnvelope creates a GitHub Pull Request Approval attestation envelope
+func createGitHubPullRequestApprovalAttestationEnvelope(t *testing.T, refName, fromID, toID string, approvers, dismissedApprovers []string) *sslibdsse.Envelope {
```
Comment (patzielinski) [inline internal/attestations/github_test.go:236]: Are there any other functions which need to change based on the addition of `dismissedApprovers` here?

File: internal/attestations/github_test.go [hunk 7]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+	t.Helper()
```
Comment (patzielinski) [inline internal/attestations/github_test.go:237]:
```suggestion
	t.Helper()
	
```

File: internal/attestations/common/common.go [hunk 8]
```diff
@@ -22,3 +22,15 @@ func PredicateToPBStruct(predicate any) (*structpb.Struct, error) {
...
+func UnmarshalPBStruct(pbStruct *structpb.Struct, out any) error {
+	// Convert structpb.Struct to JSON bytes
+	jsonBytes, err := pbStruct.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal into the provided Go struct
+	return json.Unmarshal(jsonBytes, out)
+}
```
Comment (patzielinski) [inline internal/attestations/common/common.go:36]: Would it be possible to use `AsMap` instead of having this unmarshaller?

e.g. https://github.com/gittuf/gittuf/blob/main/internal/attestations/github/v01/approval_test.go#L48

**Review by patzielinski — CHANGES_REQUESTED** [review]
Hi @DevAnuragT, thanks for the PR. A few suggestions. I also see that the coverage went down after this PR, could you please look into that?

Thanks!

File: internal/attestations/github_test.go [hunk 9]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
 
-	authorization, err := NewGitHubPullRequestApprovalAttestation(refName, fromID, toID, approvers, nil)
+	// Retrieve the stored attestation
+	retrieved, err := attestations.GetGitHubPullRequestApprovalAttestationFor(repo, appName, testRef, testID, testID)
+	assert.Nil(t, err)
+
+	// Extract the statement from the envelope
+	var statement ita.Statement
+	payloadBytes, err := base64.StdEncoding.DecodeString(retrieved.Payload)
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (DevAnuragT) [reply internal/attestations/github_test.go:125]: Thanks for the suggestion! I wanted to mention that in Go testing, the idiomatic way to assert an error is nil is using require.NoError(t, err). It’s more descriptive and provides clearer failure messages compared to require.Nil(t, err), which is more generic.

For consistency and clarity, I’ve used require.NoError throughout the tests.

Let me know if you still want to change it.

File: internal/attestations/github_test.go (hunk 6 above)
Comment (DevAnuragT) [reply internal/attestations/github_test.go:236]: Yes I have already taken care of that.

File: internal/attestations/github_test.go [hunk 10]
```diff
@@ -84,10 +90,153 @@ func TestGetGitHubPullRequestApprovalAttestation(t *testing.T) {
...
+	}
+	return result
+}
+
+/*
+TestNewPullRequestApprovalAttestation_ErrorOnEmptyLists verifies that
+NewGitHubPullRequestApprovalAttestation returns an error when both
+approvers and dismissedApprovers are empty.
+*/
+func TestNewPullRequestApprovalAttestation_ErrorOnEmptyLists(t *testing.T) {
```
Comment (adityasaky) [inline internal/attestations/github_test.go:164]: I'd skip the `_...` bit and just make that a `t.Run` description.

Comment (DevAnuragT) [reply internal/attestations/github_test.go]: Yes, I have changed it in my latest commit

File: internal/attestations/github_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (patzielinski) [reply internal/attestations/github_test.go]: Thanks for the suggestion; I figure though given our project's standard of using `require.Nil` means that it's probably better for now to change this to prevent fragmentation in style.

[the remark this answers is not in the document]
Comment (DevAnuragT) [reply internal/attestations/github_test.go]: Sure, adding a new commit.

**Comment by DevAnuragT** [comment]
I have made the necessary changes, can this be merged now?
